churaverse-plugins-client/src/gamePlugin/domain/baseGamePlugin.ts

```diff
@@ -0,0 +1,167 @@
+import { BasePlugin, IMainScene, Store, IEventBus } from 'churaverse-engine-client'
+import { GameIds } from '../interface/gameIds'
+import { GameAbortEvent } from '../event/gameAbortEvent'
+import { GameEndEvent } from '../event/gameEndEvent'
+import { UpdateGameParticipantEvent } from '../event/updateGameParticipantEvent'
+import { PriorGameDataEvent } from '../event/priorGameDataEvent'
+import { GamePluginStore } from '../store/defGamePluginStore'
+
+export abstract class BaseGamePlugin extends BasePlugin<IMainScene> {
+  /** ゲーム一意のid */
+  protected abstract readonly gameId: GameIds
+  /** ゲーム名 */
+  protected abstract readonly gameName: string
+  /** ゲームの状態を保存する変数 */
+  private _isActive: boolean
+  /** ゲームを開始したプレイヤーid */
+  private _gameOwnerId: string | undefined
+  /** ゲーム参加者のプレイヤーid */
+  private _participantIds: string[]
+  /** 自プレイヤーが途中参加かどうかを示すフラグ */
+  private _isOwnPlayerMidwayParticipant: boolean
+
+  public constructor(store: Store<IMainScene>, bus: IEventBus<IMainScene>, sceneName: IMainScene['sceneName']) {
+    super(store, bus, sceneName)
+    this._isActive = false
+    this._gameOwnerId = undefined
+    this._participantIds = []
+    this._isOwnPlayerMidwayParticipant = false
+  }
+
+  /**
+   * ゲームの状態を取得する
+   */
+  protected get isActive(): boolean {
+    return this._isActive
+  }
+
+  /**
+   * ゲームを開始したプレイヤーidを取得する
+   */
+  protected get gameOwnerId(): string | undefined {
+    return this._gameOwnerId
+  }
+
+  /**
+   * ゲーム参加者のプレイヤーidリストを取得する
+   */
+  protected get participantIds(): string[] {
+    return this._participantIds
+  }
+
+  /**
+   * 自プレイヤーが途中参加かの有無を取得する
+   */
+  protected get isOwnPlayerMidwayParticipant(): boolean {
+    return this._isOwnPlayerMidwayParticipant
+  }
+
+  public listenEvent(): void {
+    this.bus.subscribeEvent('priorGameData', this.priorGameData.bind(this))
+  }
+
+  /**
+   * ゲーム開始時に共通して登録されるイベントリスナー
+   */
+  protected subscribeGameStartEvent(): void {
+    this.bus.subscribeEvent('gameAbort', this.gameAbort)
+    this.bus.subscribeEvent('gameEnd', this.gameEnd)
+    this.bus.subscribeEvent('updateGameParticipant', this.updateGameParticipant)
+  }
+
+  /**
+   * ゲームが中断・終了時に共通して削除されるイベントリスナー
+   */
+  protected unsubscribeGameTerminationEvent(): void {
+    this.bus.unsubscribeEvent('gameAbort', this.gameAbort)
+    this.bus.unsubscribeEvent('gameEnd', this.gameEnd)
+    this.bus.unsubscribeEvent('updateGameParticipant', this.updateGameParticipant)
+  }
+
+  /**
+   * プレイヤーが参加した時に、ゲームが開始されているかを確認する。開始されている場合、必要な処理を実行する
+   */
+  private priorGameData(ev: PriorGameDataEvent): void {
+    if (ev.runningGameId !== this.gameId || this.isActive) return
+    this._isActive = true
+    this._isOwnPlayerMidwayParticipant = true
```

Comment:
isActiveを正しい値に変えるのをまずやったほうがわかりやすくなりそうです

```suggestion
    this._isActive = ev.runningGameId === this.gameId
    
    if (this.isActive) {
      this._isOwnPlayerMidwayParticipant = true
      const gamePluginStore = this.store.of('gamePlugin')
      gamePluginStore.gameLogRenderer.gameLog(`${this.gameName}が開始されています。`, 400)
      this.handleMidwayParticipant()
    }
```

Reply:
レビューありがとうございます！以下のようにプログラムを修正しました。

`if (ev.runningGameId !== this.gameId || this.isActive) return`の処理に関してですが、当初は、backendの方でメッセージのdestを`allClients`としていたのですが、今回の修正で`onlySelf`に修正しました。私の勘違いで特定のプレイヤーにメッセージを送信することができないと誤認識していました。ご確認よろしくお願いします。

```suggestion
    this._isActive = ev.runningGameId === this.gameId
    this._isOwnPlayerMidwayParticipant = true
    const gamePluginStore = this.store.of('gamePlugin')
    gamePluginStore.gameLogRenderer.gameLog(`${this.gameName}が開始されています。`, 400)
    this.handleMidwayParticipant()
```

Reply:
横から失礼します。
`this._isActive = false`なら以降の行を実行する必要ないと思うのですがどうですか？

Reply:
そうだね
`this._isActive = false`以降の処理はすでにゲームが開始していた場合の途中参加者に対する処理だから、isActiveの時だけ実行される形が正しそう

なんなら↓の部分は`handleMidwayParticipant()`内に移動してしまった方が良さそうかな
```ts
this._isOwnPlayerMidwayParticipant = true
const gamePluginStore = this.store.of('gamePlugin')
gamePluginStore.gameLogRenderer.gameLog(`${this.gameName}が開始されています。`, 400)
```

Reply:
レビューありがとうございます。
指摘されたように、`this._isActive = false`でも途中参加の処理が実行されてしまうと、異なるGameIdでも途中参加の処理が実行されるので、修正しました。


また、
```
this._isOwnPlayerMidwayParticipant = true
const gamePluginStore = this.store.of('gamePlugin')
gamePluginStore.gameLogRenderer.gameLog(`${this.gameName}が開始されています。`, 400)
```
の処理は、以下の理由から`handleMidwayParticipant`に入れずに修正しています。
- 途中参加の処理はゲーム全体で共通できるものであると考えるため。
- `priorGameData`関数のコメントに「開始されている場合、必要な処理を実行する」と明記しているので、確実に実行されるようにした方がいいと考えたため。